### DISCUSSION TITLE
How to use mime_type in derivatives block?

### DISCUSSION THREAD

**Discussion by jeremyhaile** [shrinerb/shrine#667] [Q&A] [answered]
I'd like access to the derived mime_type (or all of the metadata) when creating derivatives, but the derivatives block only takes in "original" (a reference to the original file)

```
  Attacher.derivatives do |original|
      # I want to use mime_type etc. here
   end
```

Is there any way to access that there? I tried using `original.path` and just looking at file extensions, but that is error-prone and also sometimes doesn't work (because sometimes the file I'm processing from another service doesn't even have a file extension)

The specific problems I'm trying to handle are:
1. I allow uploading images or videos...and the ImageProcessing doesn't work on videos - so I want to skip them
2. I'm getting the following error sometimes when uploading gif images, so I'm also trying to skip processing them: `Source format is multi-layer, but destination format is single-layer.`  (this is happening when I call `magick.coalesce.resize_to_limit!(128, 128)`)

My full code right now looks like this:

```
  Attacher.derivatives do |original|
    if /\.(mp4|m4v|gif)/i.match?(original.path)
      {}
    else
      magick = ImageProcessing::MiniMagick
        .source(original)
        .saver(quality: 60)

      {
        thumbnail: magick.coalesce.resize_to_limit!(128, 128),
        optimized: magick.coalesce.resize_to_limit!(1024, 1024)
      }
    end
  end
```

**Comment by janko** [comment, marked as answer]
You can pass `download: false` to the `Attacher.derivatives` call, and download the uploaded file yourself:

```rb
Attacher.derivatives(download: false) do |source|
  next {} if %w[video/mpeg image/gif ...].include?(source.mime_type)
  shrine_class.with_file(source) do |file|
    magick = ImageProcessing::MiniMagick
      .source(file)
      .saver(quality: 60)

    {
      thumbnail: magick.coalesce.resize_to_limit!(128, 128),
      optimized: magick.coalesce.resize_to_limit!(1024, 1024)
    }
  end
end
```
